File: wagerproof-ios-native/Wagerproof/Features/Auth/SignupView.swift
// SignupView.swift
//
// Create-account form with email + password + confirm, plus Apple/Google
// social fallback. Wired to `AuthStore.signUp(email:password:)`. After success
// the auth state listener routes the user; on email-verification we show a
// "check your email" message and bounce back to login after 3s.
//
// Visual: shared pixel-glyph gate background, left-aligned minimalist logo +
// header, Liquid Glass input rows + CTA, and the shared Apple/Google sign-in
// pills (matching the welcome gate).

import SwiftUI
import AuthenticationServices
import WagerproofDesign
import WagerproofStores

struct SignupView: View {
    @Environment(AuthStore.self) private var authStore
    @Environment(\.dismiss) private var dismiss

    @State private var email: String = ""
    @State private var password: String = ""
    @State private var confirmPassword: String = ""
    @State private var loading: Bool = false
    @State private var isPasswordVisible: Bool = false
    @State private var isConfirmVisible: Bool = false
    @State private var errorMessage: String?
    @State private var successMessage: String?
    @State private var currentNonce: String?
    @State private var googleLoading: Bool = false
    @State private var appleLoading: Bool = false
    @State private var appleCoordinator = AppleSignInCoordinator()
    @FocusState private var focused: Field?

    enum Field { case email, password, confirm }

    private var anyLoading: Bool { loading || googleLoading || appleLoading }
    private var formIncomplete: Bool {
        loading || email.isEmpty || password.isEmpty || confirmPassword.isEmpty || successMessage != nil
    }

    var body: some View {
        ZStack {
            AuthGateBackground()

            ScrollView {
                VStack(alignment: .leading, spacing: 0) {
                    topBar.padding(.bottom, 28)

                    Image("WagerproofLogo")
                        .resizable()
                        .scaledToFit()
                        .frame(width: 40, height: 40)
                        .padding(.bottom, 18)
                        .accessibilityHidden(true)

                    header.padding(.bottom, 28)
                    form
                }
                .padding(.horizontal, 24)
                .padding(.top, 8)
                .padding(.bottom, 24)
            }
            .scrollDismissesKeyboard(.interactively)
            .scrollIndicators(.hidden)
        }
        .preferredColorScheme(.dark)
        .navigationBarBackButtonHidden()
        .toolbar(.hidden, for: .navigationBar)
        .sensoryFeedback(.success, trigger: successMessage)
        .sensoryFeedback(.error, trigger: errorMessage)
        .sensoryFeedback(.selection, trigger: isPasswordVisible)
        .sensoryFeedback(.selection, trigger: isConfirmVisible)
    }

    // MARK: - Sub-views

    private var backButton: some View {
        Button {
            dismiss()
        } label: {
            Image(systemName: "chevron.left")
                .font(.system(size: 17, weight: .semibold))
                .foregroundStyle(.white)
                .frame(width: 40, height: 40)
                .liquidGlassBackground(in: Circle())
        }
        .buttonStyle(.plain)
        // Visual stays 40pt; tap area expands to 44pt (matches RN hitSlop).
        .contentShape(Rectangle())
        .frame(minWidth: 44, minHeight: 44)
        .disabled(loading)
        .accessibilityLabel("Back")
    }

    private var header: some View {
        VStack(alignment: .leading, spacing: 4) {
            Text("Create account")
                .font(.system(size: 26, weight: .semibold))
                .foregroundStyle(.white)
            Text("Get started with professional sports analytics")
                .font(.system(size: 15))
                .foregroundStyle(.white.opacity(0.6))
        }
    }

    private var form: some View {
        VStack(alignment: .leading, spacing: 16) {
            // Email
            AuthFieldRow(label: "Email", icon: "envelope") {
                TextField("", text: $email, prompt:
                    Text("you@example.com").foregroundStyle(.white.opacity(0.3)))
                    .keyboardType(.emailAddress)
                    .textContentType(.emailAddress)
                    .textInputAutocapitalization(.never)
                    .autocorrectionDisabled()
                    .submitLabel(.next)
                    .focused($focused, equals: .email)
                    .foregroundStyle(.white)
                    .tint(.appPrimary)
                    .onChange(of: email) { _, _ in clearMessages() }
                    .onSubmit { focused = .password }
                    .disabled(loading || successMessage != nil)
            }

            // Password
            AuthFieldRow(label: "Password", icon: "lock") {
                Group {
                    if isPasswordVisible {
                        TextField("", text: $password, prompt:
                            Text("At least 8 characters").foregroundStyle(.white.opacity(0.3)))
                            .textContentType(.newPassword)
                            .textInputAutocapitalization(.never)
                            .autocorrectionDisabled()
                    } else {
                        SecureField("", text: $password, prompt:
                            Text("At least 8 characters").foregroundStyle(.white.opacity(0.3)))
                            .textContentType(.newPassword)
                    }
                }
                .submitLabel(.next)
                .focused($focused, equals: .password)
                .foregroundStyle(.white)
                .tint(.appPrimary)
                .onChange(of: password) { _, _ in clearMessages() }
                .onSubmit { focused = .confirm }
                .disabled(loading || successMessage != nil)
            } trailing: {
                visibilityToggle(isOn: $isPasswordVisible)
            }

            // Confirm password
            AuthFieldRow(label: "Confirm Password", icon: "lock.shield") {
                Group {
                    if isConfirmVisible {
                        TextField("", text: $confirmPassword, prompt:
                            Text("Re-enter your password").foregroundStyle(.white.opacity(0.3)))
                            .textContentType(.newPassword)
                            .textInputAutocapitalization(.never)
                            .autocorrectionDisabled()
                    } else {
                        SecureField("", text: $confirmPassword, prompt:
                            Text("Re-enter your password").foregroundStyle(.white.opacity(0.3)))
                            .textContentType(.newPassword)
                    }
                }
                .submitLabel(.go)
                .focused($focused, equals: .confirm)
                .foregroundStyle(.white)
                .tint(.appPrimary)
                .onChange(of: confirmPassword) { _, _ in clearMessages() }
                .onSubmit { Task { await signUp() } }
                .disabled(loading || successMessage != nil)
            } trailing: {
                visibilityToggle(isOn: $isConfirmVisible)
            }

            // Disclaimer
            HStack(alignment: .top, spacing: 8) {
                Image(systemName: "info.circle")
                    .font(.system(size: 13))
                    .foregroundStyle(.white.opacity(0.4))
                Text("By signing up, you confirm that you are 18+ and understand this platform is for analytics only.")
                    .font(.system(size: 12))
                    .foregroundStyle(.white.opacity(0.4))
                    .lineSpacing(2)
            }
            .padding(.top, 2)

            if let error = errorMessage {
                AuthErrorBanner(message: error).transition(.opacity)
            }
            if let success = successMessage {
                AuthSuccessBanner(message: success).transition(.opacity)
            }

            // Create Account CTA
            LiquidGlassPillButton(
                title: "Create Account",
                loading: loading,
                isEnabled: !formIncomplete
            ) {
                Task { await signUp() }
            }
            .padding(.top, 4)

            divider

            // Social fallback — same pills as the welcome gate.
            appleButton
            googleButton
        }
        .animation(.easeInOut(duration: 0.2), value: errorMessage)
        .animation(.easeInOut(duration: 0.2), value: successMessage)
    }

    private var divider: some View {
        HStack(spacing: 12) {
            Rectangle().fill(.white.opacity(0.12)).frame(height: 1)
            Text("or continue with")
                .font(.system(size: 13))
                .foregroundStyle(.white.opacity(0.4))
            Rectangle().fill(.white.opacity(0.12)).frame(height: 1)
        }
        .padding(.vertical, 4)
    }

    private var appleButton: some View {
        Button {
            startAppleSignIn()
        } label: {
            AuthPillLabel(title: "Continue with Apple", style: .light, loading: appleLoading) {
                Image(systemName: "applelogo")
                    .font(.system(size: 15, weight: .medium))
                    .foregroundStyle(.black)
            }
        }
        .buttonStyle(PillPressStyle())
        .disabled(anyLoading || successMessage != nil)
        .opacity((anyLoading && !appleLoading) || successMessage != nil ? 0.6 : 1)
    }

    private var googleButton: some View {
        Button {
            Task {
                clearMessages()
                googleLoading = true
                await authStore.signInWithGoogle()
                googleLoading = false
                if let raw = authStore.lastError {
                    errorMessage = raw
                    authStore.clearError()
                }
            }
        } label: {
            AuthPillLabel(title: "Continue with Google", style: .dark, loading: googleLoading) {
                Text("G")
                    .font(.system(size: 15, weight: .heavy, design: .rounded))
                    .foregroundStyle(.white)
            }
        }
        .buttonStyle(PillPressStyle())
        .disabled(anyLoading || successMessage != nil)
        .opacity((anyLoading && !googleLoading) || successMessage != nil ? 0.6 : 1)
    }

    /// Back button on the left and the "Already have an account? Sign In" link
    /// on the right — kept at the top so users who meant to sign in don't have
    /// to scroll past the whole form to find it.
    private var topBar: some View {
        HStack(spacing: 8) {
            backButton
            Spacer(minLength: 8)
            HStack(spacing: 4) {
                Text("Already have an account?")
                    .font(.system(size: 13))
                    .foregroundStyle(.white.opacity(0.5))
                Button {
                    dismiss()
                } label: {
                    Text("Sign In")
                        .font(.system(size: 13, weight: .semibold))
                        .foregroundStyle(Color.appPrimary)
                }
                .buttonStyle(.plain)
                .disabled(loading)
            }
        }
    }

    private func visibilityToggle(isOn: Binding<Bool>) -> some View {
        Button {
            isOn.wrappedValue.toggle()
        } label: {
            Image(systemName: isOn.wrappedValue ? "eye.slash" : "eye")
                .font(.system(size: 17))
                .foregroundStyle(.white.opacity(0.45))
                .frame(width: 32, height: 32)
        }
        .buttonStyle(.plain)
        // Visual stays 32pt; tap area expands to 44pt (matches RN hitSlop).
        .contentShape(Rectangle())
        .frame(minWidth: 44, minHeight: 44)
        .accessibilityLabel(isOn.wrappedValue ? "Hide password" : "Show password")
    }

    // MARK: - State helpers

    private func clearMessages() {
        errorMessage = nil
        successMessage = nil
    }

    // MARK: - Actions

    private func validate() -> Bool {
        if email.trimmingCharacters(in: .whitespaces).isEmpty {
            errorMessage = "Please enter your email"; return false
        }
        if !email.contains("@") {
            errorMessage = "Please enter a valid email"; return false
        }
        if password.isEmpty {
            errorMessage = "Please enter a password"; return false
        }
        if password.count < 8 {
            errorMessage = "Password must be at least 8 characters"; return false
        }
        if password != confirmPassword {
            errorMessage = "Passwords do not match"; return false
        }
        return true
    }

    private func signUp() async {
        clearMessages()
        guard validate() else { return }
        loading = true
        await authStore.signUp(
            email: email.trimmingCharacters(in: .whitespaces),
            password: password
        )
        loading = false

        if let raw = authStore.lastError {
            errorMessage = Self.classify(raw: raw)
            authStore.clearError()
            return
        }

        // RN behaviour: if a session was created (auto-confirm), show "Setting
        // up your profile…" and let the auth listener route. Otherwise show
        // "check your email" and bounce back to login after 3s.
        if case .authenticated = authStore.phase {
            successMessage = "Account created! Setting up your profile..."
        } else {
            successMessage = "Account created! Please check your email to verify your account."
            email = ""; password = ""; confirmPassword = ""
            Task {
                try? await Task.sleep(nanoseconds: 3_000_000_000)
                await MainActor.run { dismiss() }
            }
        }
    }

    /// Mirrors the RN classification:
    ///   "already registered" → "An account with this email already exists"
    static func classify(raw: String) -> String {
        if raw.localizedCaseInsensitiveContains("already registered") {
            return "An account with this email already exists"
        }
        return raw
    }

    // MARK: - Apple plumbing (programmatic)

    private func startAppleSignIn() {
        clearMessages()
        let nonce = AppleNonce.random()
        currentNonce = nonce
        appleLoading = true
        appleCoordinator.start(nonceSHA256: AppleNonce.sha256Hex(nonce)) { result in
            Task { @MainActor in handleAppleCompletion(result) }
        }
    }

    private func handleAppleCompletion(_ result: Result<ASAuthorization, Error>) {
        appleLoading = false
        switch result {
        case .success(let auth):
            guard
                let credential = auth.credential as? ASAuthorizationAppleIDCredential,
                let tokenData = credential.identityToken,
                let idToken = String(data: tokenData, encoding: .utf8)
            else {
                errorMessage = "Apple sign-in returned no identity token."
                return
            }
            let nonce = currentNonce
            currentNonce = nil
            Task {
                await authStore.signInWithApple(idToken: idToken, nonce: nonce)
                if let raw = authStore.lastError {
                    errorMessage = raw
                    authStore.clearError()
                }
            }
        case .failure(let error):
            currentNonce = nil
            if let asError = error as? ASAuthorizationError, asError.code == .canceled { return }
            errorMessage = error.localizedDescription
        }
    }
}

// MARK: - Success banner

struct AuthSuccessBanner: View {
    let message: String

    var body: some View {
        HStack(spacing: 10) {
            Image(systemName: "checkmark.circle")
                .font(.system(size: 18))
                .foregroundStyle(Color.appPrimary)
            Text(message)
                .font(.system(size: 14, weight: .medium))
                .foregroundStyle(Color.appPrimary)
                .frame(maxWidth: .infinity, alignment: .leading)
        }
        .padding(14)
        .background(
            RoundedRectangle(cornerRadius: 12)
                .fill(Color.appPrimary.opacity(0.12))
        )
    }
}

File: wagerproof-android-native/app/src/main/java/com/wagerproof/app/features/auth/SignupView.kt
package com.wagerproof.app.features.auth

import androidx.activity.compose.LocalActivity
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.safeDrawingPadding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.hapticfeedback.HapticFeedbackType
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.wagerproof.app.di.appGraph
import com.wagerproof.app.features.auth.Components.AuthBackButton
import com.wagerproof.app.features.auth.Components.AuthErrorBanner
import com.wagerproof.app.features.auth.Components.AuthFieldRow
import com.wagerproof.app.features.auth.Components.AuthLogo
import com.wagerproof.app.features.auth.Components.AuthPillButton
import com.wagerproof.app.features.auth.Components.AuthPillStyle
import com.wagerproof.app.features.auth.Components.AuthSuccessBanner
import com.wagerproof.app.features.auth.Components.AuthTextField
import com.wagerproof.app.features.auth.Components.LiquidGlassPillButton
import com.wagerproof.app.features.auth.Components.PasswordVisibilityToggle
import com.wagerproof.core.design.icons.AppIcon
import com.wagerproof.core.design.tokens.AppColors
import com.wagerproof.core.services.GoogleSignInHelper
import com.wagerproof.core.stores.AuthStore
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/**
 * Create-account form — port of iOS `SignupView`. Wired to `AuthStore.signUp`.
 * On auto-confirm the auth listener routes; otherwise a "check your email"
 * banner shows and we bounce back to login after 3s.
 *
 * FIDELITY-WAIVER #201: Apple Sign-In dropped; only the Google social pill shows.
 */
@Composable
fun SignupView(
    onBack: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val authStore = appGraph().auth
    val activity = LocalActivity.current
    val scope = rememberCoroutineScope()
    val haptics = LocalHapticFeedback.current

    var email by remember { mutableStateOf("") }
    var password by remember { mutableStateOf("") }
    var confirmPassword by remember { mutableStateOf("") }
    var loading by remember { mutableStateOf(false) }
    var googleLoading by remember { mutableStateOf(false) }
    var passwordVisible by remember { mutableStateOf(false) }
    var confirmVisible by remember { mutableStateOf(false) }
    var errorMessage by remember { mutableStateOf<String?>(null) }
    var successMessage by remember { mutableStateOf<String?>(null) }

    val anyLoading = loading || googleLoading
    val formIncomplete = loading || email.isEmpty() || password.isEmpty() ||
        confirmPassword.isEmpty() || successMessage != null

    LaunchedEffect(errorMessage) {
        if (errorMessage != null) haptics.performHapticFeedback(HapticFeedbackType.LongPress)
    }
    LaunchedEffect(successMessage) {
        if (successMessage != null) haptics.performHapticFeedback(HapticFeedbackType.LongPress)
    }
    LaunchedEffect(passwordVisible, confirmVisible) {
        haptics.performHapticFeedback(HapticFeedbackType.TextHandleMove)
    }

    fun clearMessages() { errorMessage = null; successMessage = null }

    fun submit() {
        clearMessages()
        val trimmed = email.trim()
        when {
            trimmed.isEmpty() -> { errorMessage = "Please enter your email"; return }
            !trimmed.contains("@") -> { errorMessage = "Please enter a valid email"; return }
            password.isEmpty() -> { errorMessage = "Please enter a password"; return }
            password.length < 8 -> { errorMessage = "Password must be at least 8 characters"; return }
            password != confirmPassword -> { errorMessage = "Passwords do not match"; return }
        }
        scope.launch {
            loading = true
            authStore.signUp(trimmed, password)
            loading = false
            authStore.lastError?.let { raw ->
                errorMessage = classifySignupError(raw)
                authStore.clearError()
                return@launch
            }
            // Session created (auto-confirm) → success banner, listener routes.
            // Otherwise → "check your email", clear fields, bounce back after 3s.
            if (authStore.phase is AuthStore.Phase.Authenticated) {
                successMessage = "Account created! Setting up your profile..."
            } else {
                successMessage = "Account created! Please check your email to verify your account."
                email = ""; password = ""; confirmPassword = ""
                delay(3_000)
                onBack()
            }
        }
    }

    Column(
        modifier = modifier
            .fillMaxSize()
            .safeDrawingPadding()
            .imePadding()
            .verticalScroll(rememberScrollState())
            .padding(horizontal = 24.dp)
            .padding(top = 8.dp, bottom = 24.dp),
    ) {
        // Top bar: back + "Already have an account? Sign In".
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(bottom = 28.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            AuthBackButton(onClick = onBack, enabled = !loading)
            Box(Modifier.weight(1f))
            Text("Already have an account?", color = Color.White.copy(alpha = 0.5f), fontSize = 13.sp)
            Box(Modifier.width(4.dp))
            Text(
                text = "Sign In",
                color = AppColors.appPrimary,
                fontSize = 13.sp,
                fontWeight = FontWeight.SemiBold,
                modifier = Modifier.clickableNoRipple(enabled = !loading, onClick = onBack),
            )
        }

        AuthLogo(modifier = Modifier.padding(bottom = 18.dp))

        Column(Modifier.padding(bottom = 28.dp), verticalArrangement = Arrangement.spacedBy(4.dp)) {
            Text("Create account", color = Color.White, fontSize = 26.sp, fontWeight = FontWeight.SemiBold)
            Text(
                "Get started with professional sports analytics",
                color = Color.White.copy(alpha = 0.6f),
                fontSize = 15.sp,
            )
        }

        Column(verticalArrangement = Arrangement.spacedBy(16.dp)) {
            AuthFieldRow(label = "Email", icon = "envelope.fill") {
                AuthTextField(
                    value = email,
                    onValueChange = { email = it; clearMessages() },
                    placeholder = "you@example.com",
                    keyboardType = KeyboardType.Email,
                    imeAction = ImeAction.Next,
                    enabled = !loading && successMessage == null,
                )
            }

            AuthFieldRow(
                label = "Password",
                icon = "lock.fill",
                trailing = { PasswordVisibilityToggle(passwordVisible) { passwordVisible = !passwordVisible } },
            ) {
                AuthTextField(
                    value = password,
                    onValueChange = { password = it; clearMessages() },
                    placeholder = "At least 8 characters",
                    isPassword = true,
                    passwordVisible = passwordVisible,
                    imeAction = ImeAction.Next,
                    enabled = !loading && successMessage == null,
                )
            }

            AuthFieldRow(
                label = "Confirm Password",
                icon = "lock.shield.fill",
                trailing = { PasswordVisibilityToggle(confirmVisible) { confirmVisible = !confirmVisible } },
            ) {
                AuthTextField(
                    value = confirmPassword,
                    onValueChange = { confirmPassword = it; clearMessages() },
                    placeholder = "Re-enter your password",
                    isPassword = true,
                    passwordVisible = confirmVisible,
                    imeAction = ImeAction.Go,
                    keyboardActions = KeyboardActions(onGo = { submit() }),
                    enabled = !loading && successMessage == null,
                )
            }

            // 18+ / analytics-only disclaimer.
            Row(
                modifier = Modifier.padding(top = 2.dp),
                horizontalArrangement = Arrangement.spacedBy(8.dp),
            ) {
                AppIcon.fromSystemName("info.circle")?.imageVector?.let {
                    Icon(it, contentDescription = null, tint = Color.White.copy(alpha = 0.4f), modifier = Modifier.padding(top = 1.dp))
                }
                Text(
                    "By signing up, you confirm that you are 18+ and understand this platform is for analytics only.",
                    color = Color.White.copy(alpha = 0.4f),
                    fontSize = 12.sp,
                )
            }

            errorMessage?.let { AuthErrorBanner(it) }
            successMessage?.let { AuthSuccessBanner(it) }

            LiquidGlassPillButton(
                title = "Create Account",
                loading = loading,
                isEnabled = !formIncomplete,
                onClick = { submit() },
                modifier = Modifier.padding(top = 4.dp),
            )

            // "or continue with" divider.
            Row(
                modifier = Modifier.padding(vertical = 4.dp),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(12.dp),
            ) {
                Box(Modifier.weight(1f).height(1.dp).background(Color.White.copy(alpha = 0.12f)))
                Text("or continue with", color = Color.White.copy(alpha = 0.4f), fontSize = 13.sp)
                Box(Modifier.weight(1f).height(1.dp).background(Color.White.copy(alpha = 0.12f)))
            }

            AuthPillButton(
                title = "Continue with Google",
                style = AuthPillStyle.Dark,
                loading = googleLoading,
                enabled = !anyLoading && successMessage == null,
                onClick = {
                    val act = activity ?: return@AuthPillButton
                    scope.launch {
                        clearMessages()
                        googleLoading = true
                        try {
                            val idToken = GoogleSignInHelper(act).signIn()
                            authStore.signInWithGoogle(idToken)
                            authStore.lastError?.let { raw -> errorMessage = raw; authStore.clearError() }
                        } catch (t: Throwable) {
                            if (!GoogleSignInHelper(act).isUserCancellation(t)) {
                                errorMessage = t.message ?: "Google sign-in failed."
                            }
                        } finally {
                            googleLoading = false
                        }
                    }
                },
            ) {
                Text(text = "G", color = Color.White, fontSize = 15.sp, fontWeight = FontWeight.Black)
            }
        }
    }
}

/** Mirrors the iOS/RN classification: "already registered" → friendly copy. */
internal fun classifySignupError(raw: String): String = when {
    raw.contains("already registered", ignoreCase = true) ->
        "An account with this email already exists"
    else -> raw
}
